src/ioctls/vm.rs

```diff
@@ -1258,62 +1258,89 @@ mod tests {
         let evtfd2 = EventFd::new(EFD_NONBLOCK).unwrap();
         let evtfd3 = EventFd::new(EFD_NONBLOCK).unwrap();
-        if cfg!(any(target_arch = "x86", target_arch = "x86_64")) {
-            assert!(vm_fd.create_irq_chip().is_ok());
-            assert!(vm_fd.register_irqfd(&evtfd1, 4).is_ok());
-            assert!(vm_fd.register_irqfd(&evtfd2, 8).is_ok());
-            assert!(vm_fd.register_irqfd(&evtfd3, 4).is_ok());
-        }
 
-        // On aarch64, this fails because setting up the interrupt controller is mandatory before
-        // registering any IRQ.
+        assert!(vm_fd.create_irq_chip().is_ok());
+
+        assert!(vm_fd.register_irqfd(&evtfd1, 4).is_ok());
+        assert!(vm_fd.register_irqfd(&evtfd2, 8).is_ok());
+        assert!(vm_fd.register_irqfd(&evtfd3, 4).is_ok());
+        assert!(vm_fd.unregister_irqfd(&evtfd2, 8).is_ok());
+        // KVM irqfd doesn't report failure on this case:(
+        assert!(vm_fd.unregister_irqfd(&evtfd2, 8).is_ok());
+
+        // Duplicated eventfd registration.
         // On x86_64 this fails as the event fd was already matched with a GSI.
         assert!(vm_fd.register_irqfd(&evtfd3, 4).is_err());
         assert!(vm_fd.register_irqfd(&evtfd3, 5).is_err());
+        // KVM irqfd doesn't report failure on this case:(
+        assert!(vm_fd.unregister_irqfd(&evtfd3, 5).is_ok());
     }
 
     #[test]
-    #[cfg(any(
-        target_arch = "x86",
-        target_arch = "x86_64",
-        target_arch = "arm",
-        target_arch = "aarch64"
-    ))]
-    fn test_unregister_irqfd() {
+    #[cfg(target_arch = "aarch64")]
+    fn test_register_unregister_irqfd() {
         let kvm = Kvm::new().unwrap();
         let vm_fd = kvm.create_vm().unwrap();
         let evtfd1 = EventFd::new(EFD_NONBLOCK).unwrap();
         let evtfd2 = EventFd::new(EFD_NONBLOCK).unwrap();
         let evtfd3 = EventFd::new(EFD_NONBLOCK).unwrap();
-        if cfg!(any(target_arch = "x86", target_arch = "x86_64")) {
-            assert!(vm_fd.create_irq_chip().is_ok());
-            assert!(vm_fd.register_irqfd(&evtfd1, 4).is_ok());
-            assert!(vm_fd.register_irqfd(&evtfd2, 8).is_ok());
-            assert!(vm_fd.register_irqfd(&evtfd3, 4).is_ok());
-            assert!(vm_fd.unregister_irqfd(&evtfd2, 8).is_ok());
-            // KVM irqfd doesn't report failure on this case:(
-            assert!(vm_fd.unregister_irqfd(&evtfd2, 8).is_ok());
-        }
+
+        let mut gic_device = kvm_bindings::kvm_create_device {
+            type_: kvm_device_type_KVM_DEV_TYPE_ARM_VGIC_V3,
+            fd: 0,
+            flags: 0,
+        };
+
+        let vgic_fd = match vm_fd.create_device(&mut gic_device) {
+            Ok(fd) => fd,
+            Err(_) => {
+                gic_device.type_ = kvm_device_type_KVM_DEV_TYPE_ARM_VGIC_V2;
+                vm_fd
+                    .create_device(&mut gic_device)
+                    .expect("Cannot create KVM vGIC device")
+            }
+        };
```

Comment:
We probably want to add these lines to a test function so that they can be used in both test with irqfds and the test for create_device.

We can have a function such as:

```rust
#[cfg(test)]
/// Creates a GIC device using the KVM_CREATE_DEVICE_TEST flags.
pub (crate) fn create_test_gic_device() -> Result<DeviceFd> {
    let mut gic_device = kvm_bindings::kvm_create_device {
            type_: kvm_device_type_KVM_DEV_TYPE_ARM_VGIC_V3,
            fd: 0,
            flags: KVM_CREATE_DEVICE_TEST,
        };
       let device_fd = match vm.create_device(&mut gic_device) {
            Ok(fd) => fd,
            Err(_) => {
                gic_device.type_ = kvm_device_type_KVM_DEV_TYPE_ARM_VGIC_V2;
                vm.create_device(&mut gic_device)
            }
        };
}
```

Then in this test and in `test_create_device` from device.rs we can call this function and we don't need to have duplicated code. This will require some changes to the existing tests to test the error cases. What do you think?

Reply:
Hi @andreeaflorescu ! I got the idea and it really makes sense for reducing code duplication. 

However I do have one concern: refer to the [documentation of the kvm kernel api](https://www.kernel.org/doc/Documentation/virtual/kvm/api.txt), the `KVM_CREATE_DEVICE_TEST` flag will only check only test whether the device type is supported (instead of really create a vGIC I think), and therefore in the scope of this PR, if I set this flag, the `set_device_attr` will fail, leading to the test fail. Without this vGIC setup, I don't think we can test the successful register/unregister irqfd case on aarch64. Referencing the GIC implementation in [firecracker](https://github.com/firecracker-microvm/firecracker/blob/f55da99544f16280c31919b2596522c18d83e5e1/src/arch/src/aarch64/gic.rs#L116), this flag is also not set.

Wondering if you have any comments or ideas about this problem. Thank you!

Reply:
An idea occured to me that we probably can make this flag as an input argument of this `create_test_gic_device`, what do you think?

Reply:
Sounds good!

Reply:
Hi @andreeaflorescu, just updated the PR with regard of your comments. Would you mind reviewing it again? Thanks!